packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -692,6 +701,38 @@ class TextSelectionOverlay {
 
   void _handleAnyDragEnd(DragEndDetails details) => _selectionOverlay.hideMagnifier(shouldShowToolbar: !_selection.isCollapsed);
 
+  // The center of the handle is where it's easier to drag, and
+  // [details.globalPosition] is probably close to center.
+  // Return offset from center of handle to center of correct text line.
+  // The offset use to shift real drag point on handle to text position point.
+  // Make sure even if the drag events happened on other position
+  // of handle, not center. The text position point always locate in
+  // correct text line.
+  Offset _getOffsetToTextPositionPoint(TextSelectionHandleType type) {
+    final Size handleSize = selectionControls!.getHandleSize(
+      renderObject.preferredLineHeight,
+    );
+
+    // Try to shift center of handle to top by half of handle height.
+    final double halfHandleHeight = handleSize.height / 2;
+
+    // [getHandleAnchor] use to shift selection end point to left top point
+    // of handle rect when build handle widget.
+    // End point is at bottom of selection rect, which is at the bottom of
+    // the text line too.
+    // Try to shift handle top to selection end point by dy of handle anchor.
+    final double handleAnchorDy = selectionControls!.getHandleAnchor(type, renderObject.preferredLineHeight).dy;
+
+    // Try to shift selection end point to center of correct text line
+    // by half preferred line height.
+    final double halfPreferredLineHeight = renderObject.preferredLineHeight / 2;
+
+    // When dragging handle and move, the offsetX from drag event is accurate.
+    // Only need to compute different offsetY from different handle.
```

Comment:
The x offset is accurate, so we only need to adjust the y position.

Reply:
Done.